badge-maker/lib/xml.js

```diff
@@ -0,0 +1,82 @@
+/**
+ * @module
+ */
+
+'use strict'
+
+function stripXmlWhitespace(xml) {
+  return xml.replace(/>\s+/g, '>').replace(/<\s+/g, '<').trim()
+}
+
+function escapeXml(s) {
+  if (typeof s === 'number') {
+    return s
+  } else if (s === undefined || typeof s !== 'string') {
+    return undefined
+  } else {
+    return s
+      .replace(/&/g, '&amp;')
+      .replace(/</g, '&lt;')
+      .replace(/>/g, '&gt;')
+      .replace(/"/g, '&quot;')
+      .replace(/'/g, '&apos;')
+  }
+}
+
+/**
+ * Representation of an XML element
+ */
+class XmlElement {
+  /**
+   * Xml Element Constructor
+   *
+   * @param {object} attrs Refer to individual attrs
+   * @param {string} attrs.tag
+   *    Name of the XML tag
+   * @param {Array.<string|module:badge-maker/lib/xml-element~XmlElement>} [attrs.content=[]]
+   *    Array of objects to render inside the tag. content may contain a mix of
+   *    string and XmlElement objects. If content is `[]` or ommitted the
+   *    element will be rendered as a self-closing element.
+   * @param {object} [attrs.attrs={}]
+   *    Object representing the tag's attributes as name/value pairs
+   */
+  constructor({ tag, content = [], attrs = {} }) {
+    this.tag = tag
+    this.content = content
+    this.attrs = attrs
+  }
+
+  /**
+   * Render the XML element to a string, applying appropriate escaping
+   *
+   * @returns {string} String representation of the XML element
+   */
+  render() {
+    let attrsStr = ''
+    attrsStr = Object.entries(this.attrs)
+      .map(([k, v]) => `${k}="${escapeXml(v)}"`)
+      .join(' ')
+    if (this.content.length > 0) {
+      let xmlStr = `<${this.tag}`
+      if (attrsStr === '') {
+        xmlStr += '>'
+      } else {
+        xmlStr += ` ${attrsStr}>`
+      }
+      xmlStr += this.content
+        .map(function (el) {
+          if (el instanceof XmlElement) {
+            return el.render()
+          } else {
+            return escapeXml(el)
```

Comment:
This recursive approach means `content` can be a mixed array containing strings and other `XmlElement` objects and we won't double-escape when nesting elements.